src/stampit.1
.\" Generated by scdoc 1.11.0
.\" Complete documentation for this program is not available as a GNU info page
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.\" Begin generated content:
.TH "STAMPIT" "1" "2020-11-22" "V0.3.2"
.P
.SH NAME
.P
stampit - prefix each line with an RFC 3339 compliant timestamp
.P
.SH SYNOPSIS
.P
\fBstampit\fR [\fIOPTIONS\fR] [\fITEXT\fR …]
.P
.SH DESCRIPTION
.P
\fBstampit\fR prefixes each line received on standard input with a
timestamp in the format \fIYYYY-MM-DDTHH:MM:SS.µµµµµµ±hh:mm\fR.
.P
If \fBstampit\fR receives a SIGHUP and \fIFILENAME\fR has been given, \fBstampit\fR
will reopen the file, something that can be useful when rotating logfiles
written to through  \fBstampit\fR.
.P
The optional \fITEXT\fR will be stamped and written verbatim as the first
line before \fBstampit\fR starts reading from standard input.
.P
.SH OPTIONS
.P
\fB-h\fR, \fB--help\fR
.RS 4
Print a short help text to standard error stream and exit
.P
.RE
\fB-v\fR, \fB--version\fR
.RS 4
Show version information and exit
.P
.RE
\fB-c\fR, \fB--copyright\fR
.RS 4
Show copying policy and exit
.P
.RE
\fB-l\fR, \fB--localtime\fR
.RS 4
Use local time rather than UTC
.P
.RE
\fB-o\fR \fIOUTPUT\fR, \fB--output\fR=\fIOUTPUT\fR
.RS 4
Path to output file
.P
.RE
RFC 3339 timestamps are given with µs precision.
.P
.SH EXAMPLES
.P
$ (echo A; sleep 2s; echo B; sleep 3s; echo C) | stampit
.br
2020-11-20T14:45:13.612858+00:00	A
2020-11-20T14:45:15.615734+00:00	B
2020-11-20T14:45:18.621444+00:00	C
.P
.SH AUTHOR
.P
Klaus Alexander Seistrup <klaus@seistrup.dk>
.P
.SH DEVELOPMENT
.P
Please see git repository:
.P
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.IP \(bu 4
.\}
https://github.com/kseistrup/timestamp/

.RE
.P
.SH COPYRIGHT
.P
Copyright ©2017-20, Klaus Alexander Seistrup.
.P
.SH LICENSE
.P
GNU General Public License, version 3 or later.
.P
.SH SEE ALSO
.P
\fBtimestamp\fR(1),
\fBts\fR(1) from the moreutils package,
\fBkill\fR(1) and \fBkillall\fR(1) for how to send a SIGHUP to \fBstampit\fR,
and \fBlogrotate\fR(8).


debian/changelog
timestamp (1.6-1) unstable; urgency=low

  * Fix "linefeed-is-output-when-first-char-of-next-line-is-received" buglet.

 -- Adam Sjøgren <asjo@koldfront.dk>  Sat, 22 Oct 2011 22:55:48 +0200

timestamp (1.5-1) unstable; urgency=low

  * Change optional argument [FORMAT] to optional option [-f/--format
    FORMAT], and add optional argument [LOGFILE]. This change is
    incompatible with version 1.4 of timestamp.
  * Reopen LOGFILE when SIGHUP is received, if LOGFILE was given (making
    it possible to do a killall -HUP timestamp after rotating logfiles
    written to via timestamp LOGFILE).

 -- Adam Sjøgren <asjo@koldfront.dk>  Tue, 23 Aug 2011 22:57:05 +0200

timestamp (1.4-3) unstable; urgency=low

  * Really add the GPLv3 license this time

 -- Klaus Alexander Seistrup <klaus@seistrup.dk>  Tue, 15 Mar 2011 20:37:40 +0100

timestamp (1.4-2) unstable; urgency=low

  * Actually added the GPLv3 license

 -- Klaus Alexander Seistrup <klaus@seistrup.dk>  Tue, 15 Mar 2011 20:29:20 +0100

timestamp (1.4-1) unstable; urgency=low

  * Now calls setlocale() at invocation time
  * GPLv2+ → GPLv3+ :)

 -- Klaus Alexander Seistrup <klaus@seistrup.dk>  Tue, 15 Mar 2011 20:25:30 +0100

timestamp (1.3-1) unstable; urgency=low

  * Output is now line buffered
  * Timestamp buffer reduced to 128 characters (was 1024)

 -- Klaus Alexander Seistrup <klaus@seistrup.dk>  Tue, 15 Mar 2011 13:34:47 +0100

timestamp (1.2-1) unstable; urgency=low

  * Now accepts optional format string

 -- Klaus Alexander Seistrup <klaus@seistrup.dk>  Tue, 15 Mar 2011 12:22:32 +0100

timestamp (1.1-1) unstable; urgency=low

  * Now checks return value of all library calls and exits in case of error

 -- Klaus Alexander Seistrup <klaus@seistrup.dk>  Mon, 14 Mar 2011 23:16:35 +0100

timestamp (1.0-1) unstable; urgency=low

  * Initial release.

 -- Adam Sjøgren <asjo@koldfront.dk>  Mon, 14 Mar 2011 21:16:35 +0100


src/stampit.c
/*****************************************************************************
* Copyright ©2017-20 Klaus Alexander Seistrup <klaus@seistrup.dk>            *
*                                                                            *
* This program is free software; you can redistribute it and/or modify it    *
* under the terms of the GNU General Public License as published by the Free *
* Software Foundation; either version 3 of the License, or (at your option)  *
* any later version.                                                         *
*                                                                            *
* This program is distributed in the hope that it will be useful, but with-  *
* out any warranty; without even the implied warranty of merchantability or  *
* fitness for a particular purpose. See the GNU General Public License for   *
* more details.  <http://gplv3.fsf.org/>                                     *
*****************************************************************************/

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <locale.h>
#include <signal.h>
#include <libgen.h>
#include <time.h>

#include "version.h"

#ifdef __USE_MISC
#define USAGE \
"Usage: %s [OPTIONS] [TEXT …]\n" \
"\n" \
"positional arguments:\n" \
"  TEXT                  optional text to timestamp\n" \
"\n" \
"optional arguments:\n" \
"  -h, --help            show this help message and exit\n" \
"  -v, --version         show version information and exit\n" \
"  -c, --copyright       show copying policy and exit\n" \
"  -l, --localtime       use local time rather than UTC\n" \
"  -o OUTPUT, --output OUTPUT\n" \
"                        path to output file\n" \
"\n" \
"If no TEXT is given on the command line, text is read\n" \
"from standard input and stamped one line at a time.\n" \
"\n" \
"RFC 3339 timestamps are given with µs precision.\n" \
"\n" \
"If stampit receives a SIGHUP and OUTPUT has been given,\n" \
"stampit will reopen the file.\n"
#else
#define USAGE \
"Usage: %s [OPTIONS] [TEXT …]\n" \
"\n" \
"positional arguments:\n" \
"  TEXT                  optional text to timestamp\n" \
"\n" \
"optional arguments:\n" \
"  -h, --help            show this help message and exit\n" \
"  -v, --version         show version information and exit\n" \
"  -c, --copyright       show copying policy and exit\n" \
"  -o OUTPUT, --output OUTPUT\n" \
"                        path to output file\n" \
"\n" \
"If no TEXT is given on the command line, text is read\n" \
"from standard input and stamped one line at a time.\n" \
"\n" \
"RFC 3339 timestamps are given in UTC with µs precision.\n"
"\n" \
"If stampit receives a SIGHUP and OUTPUT has been given,\n" \
"stampit will reopen the file.\n"
#endif

char *filename = NULL;

#ifdef __USE_MISC
int use_localtime = 0;
#endif

static void
timestamp(void)
{
   int res, usec;
   int hh = 0, mm = 0, ss = 0;
   time_t tnow = 0;
   struct timespec ts = {
      .tv_sec  = 0,
      .tv_nsec = 0
   };
   struct tm *tm = NULL;

   res = clock_gettime(CLOCK_REALTIME, &ts);

   if (res != 0)
     {
	perror("Cannot get current time");
	exit(1);
     }

   tnow = ts.tv_sec;
   usec = (int) ((ts.tv_nsec + 500) / 1000);

   while (usec > 1000000)
     {
	usec -= 1000000;
	tnow += 1;
     }

#ifdef __USE_MISC
   if (use_localtime == 1)
     {
	tm = localtime(&tnow);
	ss = tm->tm_gmtoff;
	hh =  abs(ss) / 3600;
	mm = (abs(ss) % 3600) / 60;
     }
   else
#endif
      tm = gmtime(&tnow);

   if (tm == (struct tm *) NULL)
      exit(1);

   // YYYY-MM-DDThh:mm:ss.uuuuuu+00:00t
   // ----+----|----+----|----+----|---
   // = 32 chars for the stamp and 1 char for the TAB
   res = printf("%04d-%02d-%02dT%02d:%02d:%02d.%06d%c%02d:%02d\t",
		tm->tm_year + 1900,
		tm->tm_mon + 1,
		tm->tm_mday,
		tm->tm_hour,
		tm->tm_min,
		tm->tm_sec,
		usec,
		ss < 0 ? '-' : '+', hh, mm);

   if (res != 33)
      exit(1);
}

static void
print(char *text, int end)
{
   if (fputs(text, stdout) == EOF)
      exit(1);
   if (putchar(end) == EOF)
      exit(1);
}

static void
open_out(int signum)
{
   if (filename != NULL)
     {
	if (freopen(filename, "a", stdout) == (FILE *) NULL)
	  {
	     perror(filename);
	     exit(1);
	  }
     }

   if (setvbuf(stdout, NULL, (int) _IOLBF, 0) != 0)
     {
	perror("Cannot set line buffering");
	exit(1);
     }
}

int
main(int argc, char **argv)
{
   int c, t;
   int opt, idx = 0;
   char *prog = basename(argv[0]);
#ifdef __USE_MISC
   char *short_opts = ":hvclo:";
#else
   char *short_opts = ":hvco:";
#endif

   const struct option long_opts[] = {
      {"help",      0, NULL, (int) 'h'},
      {"version",   0, NULL, (int) 'v'},
      {"copyright", 0, NULL, (int) 'c'},
      {"output",    1, NULL, (int) 'o'},
      { NULL,       0, NULL,        0 }
   };

   (void) setlocale(LC_ALL, "");

   argv[0] = prog;
   opterr = 0;

   while ((opt = getopt_long(argc, argv, short_opts, long_opts, &idx)) != EOF)
     {
	switch (opt)
	  {
	  case 'h':
	     (void) fprintf(stderr, USAGE, prog);
	     return 0;
	     /*@fallthrough@ */
	  case 'v':
	     (void) fprintf(stderr, "%s/%s\n", STAMPIT, VERSION);
	     return 0;
	     /*@fallthrough@ */
	  case 'c':
	     (void) fputs("GNU General Public License v3+\n", stderr);
	     return 0;
	     /*@fallthrough@ */
	  case 'o':
	     filename = optarg;
	     break;
#ifdef __USE_MISC
	  case 'l':
	     use_localtime = 1;
	     break;
#endif
	  default:
	     if (optopt)
	       {
		  if (opt == (int) '?')
		     (void) fprintf(stderr, "%s: Invalid option ‘%c’.\n",
				    prog, optopt);
		  else if (opt == (int) ':')
		     (void) fprintf(stderr,
				    "%s: Option ‘%c’ requires an argument.\n",
				    prog, optopt);
	       }
	     else if (optind - 1 <= argc)
		(void) fprintf(stderr, "%s: Unrecognized option ‘%s’.\n",
			       prog, argv[optind - 1]);
	     (void) fprintf(stderr,
			    "Try ‘%s --help’ for more information.\n",
			    prog);
	     return 1;
	  }
     }
   argv += optind;
   argc -= optind;

   if (signal(SIGHUP, open_out) == SIG_IGN)
      signal(SIGHUP, SIG_IGN);

   open_out(0);

   if (argc > 0)
     {
	timestamp();

	// Python: print(' '.join(argv))
	for (idx = 0; idx < (argc - 1); ++idx)
	   print(argv[idx], (int) ' ');

	print(argv[idx], (int) '\n');

	return 0;
     }

   if ((c = getchar()) == EOF)
      return 0;

   if (ungetc(c, stdin) != c)
      return 1;

   timestamp();

   for (;;)
     {
	if ((c = getchar()) == EOF)
	   return 0;

	if (c == (int) '\n')
	  {
	     if (putchar((int) '\n') == EOF)
		return 1;

	     if ((t = getchar()) == EOF)
		return 0;

	     if (ungetc(t, stdin) != t)
		return 1;

	     timestamp();
	  }
	else if (putchar(c) == EOF)
	   return 1;
     }
}

// indent -gnu -npcs -i3
